### PR TITLE
Bunch of improvements surrounding thread names and reporting of such to the profiler

### PR DESCRIPTION
**Pull request by alexrp**
Supersedes #2830.

### CODE REVIEW

File: mono/utils/mono-threads-posix.c [hunk 1]
```diff
@@ -279,7 +280,31 @@ mono_native_thread_create (MonoNativeThreadId *tid, gpointer func, gpointer arg)
...
+	static int (*pthread_setname_np_fn) (const char *);
+
+	/*
+	 * We can't set the thread name for other threads, but we can at least make
+	 * it work for threads that try to change their own name.
+	 */
+	if (tid != mono_native_thread_id_get ())
+		return;
+
+	/* The function is only available on OS X 10.6+, so find it at runtime. */
```
Comment (akoeplinger) [inline mono/utils/mono-threads-posix.c]: AFAIK we're requiring 10.7+ now, so this could be simplified.

Comment (alexrp) [inline mono/utils/mono-threads-posix.c]: Do we state this clearly somewhere? If so, I'm all for getting rid of this nonsense.

Comment (akoeplinger) [inline mono/utils/mono-threads-posix.c]: http://www.mono-project.com/docs/about-mono/supported-platforms/osx/#introduction-to-mono-on-macos-x

Comment (alexrp) [inline mono/utils/mono-threads-posix.c]: OK, will get rid of the `dlsym` stuff.

**Comment by alexrp** [comment]
Hmm, I just realized, I think this may fix the rare "missing thread Finalizer" profiler test failure that pops up on Jenkins every now and then. I think the issue fixed in 8e6b87f97ab115b7baeb81169839981aa2f8a205 was what caused that to happen randomly.

**Comment by migueldeicaza** [comment]
This is lovely!

**Pull request merged**